Sources/ComposableArchitecture/Internal/CurrentValueRelay.swift

```diff
@@ -14,42 +15,128 @@ final class CurrentValueRelay<Output>: Publisher {
+    self.lock.sync {
+      guard let index = self.subscriptions.firstIndex(of: subscription)
+      else { return }
+      self.subscriptions.remove(at: index)
     }
   }
 }
 
 extension CurrentValueRelay {
-  final class Subscription<Downstream: Subscriber>: Combine.Subscription
-  where Downstream.Input == Output, Downstream.Failure == Failure {
-    private var demandBuffer: DemandBuffer<Downstream>?
+  fileprivate final class Subscription: Combine.Subscription, Equatable {
+    private var demand = Subscribers.Demand.none
+    private var downstream: (any Subscriber<Output, Never>)?
+    private let lock: os_unfair_lock_t
+    private var receivedLastValue = false
+    private var upstream: CurrentValueRelay?
 
-    init(downstream: Downstream) {
-      self.demandBuffer = DemandBuffer(subscriber: downstream)
+    init(upstream: CurrentValueRelay, downstream: any Subscriber<Output, Never>) {
+      self.upstream = upstream
+      self.downstream = downstream
+      self.lock = os_unfair_lock_t.allocate(capacity: 1)
+      self.lock.initialize(to: os_unfair_lock())
     }
 
-    func forwardValueToBuffer(_ value: Output) {
-      _ = demandBuffer?.buffer(value: value)
+    deinit {
+      self.lock.deinitialize(count: 1)
+      self.lock.deallocate()
+    }
+
+    func cancel() {
+      self.lock.sync {
+        self.downstream = nil
+        self.upstream?.remove(self)
+        self.upstream = nil
+      }
+    }
+
+    func receive(_ value: Output) {
+      guard let downstream else { return }
+
+      switch self.demand {
+      case .unlimited:
+        // NB: Adding to unlimited demand has no effect and can be ignored.
+        _ = downstream.receive(value)
+
+      case .none:
+        self.lock.sync {
+          self.receivedLastValue = false
+        }
+
+      default:
+        self.lock.sync {
+          self.receivedLastValue = true
+          self.demand -= 1
+        }
+        let moreDemand = downstream.receive(value)
+        self.lock.sync {
+          self.demand += moreDemand
+        }
+      }
     }
 
     func request(_ demand: Subscribers.Demand) {
-      _ = demandBuffer?.demand(demand)
+      precondition(demand > 0, "Demand must be greater than zero")
+
+      guard let downstream else { return }
+
+      self.lock.lock()
+      self.demand += demand
+
+      guard
+        !self.receivedLastValue,
+        let value = self.upstream?.currentValue
+      else {
+        self.lock.unlock()
+        return
+      }
+
+      self.receivedLastValue = true
+
+      switch self.demand {
+      case .unlimited:
+        self.lock.unlock()
+        // NB: Adding to unlimited demand has no effect and can be ignored.
+        _ = downstream.receive(value)
+
+      default:
+        self.demand -= 1
+        self.lock.unlock()
+        let moreDemand = downstream.receive(value)
+        self.lock.lock()
+        self.demand += moreDemand
+        self.lock.unlock()
+      }
```

Comment:
I'm not super happy with the locking dance but this implementation seemed to run best for me. 

From my understanding the key points are:
- mutating `demand` and `receivedLastValue` needs to be locked.
- as soon as `demand` reaches `unlimited`, `receive(_:)` will always call `downstream.receive(value)`

Previously, `DemandBuffer` would handle most of this, however, it also had the potential to buffer all upstream values. In practice, this probably never happened, it would really depend on the downstream subscriber. The new behaviour only sends the current value if it hasn't already been received. This appears to be the same as `CurrentValueSubject`.